packages/next/src/client/components/segment-cache/cache.ts

```diff
@@ -0,0 +1,475 @@
+  routeKey: RouteCacheKey,
+  segmentPath: string,
+  accessToken: string | null
+): Promise<void> {
+  // This function is allowed to use async/await because it contains the actual
+  // fetch that gets issued on a cache miss. Notice though that it does not
+  // return anything; it writes the result to the cache entry directly.
+  //
+  // Segment fetches are non-blocking so we don't need to ping the scheduler
+  // on completion.
+  const href = routeKey.href
+  try {
+    const response = await fetchSegmentPrefetchResponse(
+      href,
+      accessToken === '' ? segmentPath : `${segmentPath}.${accessToken}`
+    )
+    if (!response || !response.ok || !response.body) {
+      // Server responded with an error. We should still cache the response, but
+      // we can try again after 10 seconds.
+      rejectSegmentCacheEntry(segmentCacheEntry, Date.now() + 10 * 1000)
+      return
+    }
+    // Wrap the original stream in a new stream that never closes. That way the
+    // Flight client doesn't error if there's a hanging promise. See comment in
+    // createUnclosingPrefetchStream for more details.
+    const prefetchStream = createUnclosingPrefetchStream(response.body)
+    const serverData = await (createFromNextReadableStream(
+      prefetchStream
+    ) as Promise<SegmentPrefetch>)
+    if (serverData.buildId !== getAppBuildId()) {
+      // The server build does not match the client. Treat as a 404. During
+      // an actual navigation, the router will trigger an MPA navigation.
+      // TODO: Consider moving the build ID to a response header so we can check
+      // it before decoding the response, and so there's one way of checking
+      // across all response types.
+      rejectSegmentCacheEntry(segmentCacheEntry, Date.now() + 10 * 1000)
+      return
+    }
+    fulfillSegmentCacheEntry(
+      segmentCacheEntry,
+      serverData.rsc,
+      serverData.loading,
+      // TODO: The server does not currently provide per-segment stale time.
+      // So we use the stale time of the route.
+      route.staleAt
+    )
+  } catch (error) {
+    // Either the connection itself failed, or something bad happened while
+    // decoding the response.
+    rejectSegmentCacheEntry(segmentCacheEntry, Date.now() + 10 * 1000)
+  }
+}
+
+async function fetchSegmentPrefetchResponse(
+  href: string,
+  segmentPath: string
+): Promise<Response | null> {
+  const headers: RequestHeaders = {
+    [RSC_HEADER]: '1',
+    [NEXT_ROUTER_PREFETCH_HEADER]: '1',
+    [NEXT_ROUTER_SEGMENT_PREFETCH_HEADER]: segmentPath,
+  }
+  const fetchPriority = 'low'
+  const response = await createFetch(new URL(href), headers, fetchPriority)
+  const contentType = response.headers.get('content-type')
+  const isFlightResponse =
+    contentType && contentType.startsWith(RSC_CONTENT_TYPE_HEADER)
+  if (!response.ok || !isFlightResponse) {
+    return null
+  }
+  return response
+}
+
+function createPromiseWithResolvers<T>(): PromiseWithResolvers<T> {
```

Comment:
I think you can import the [existing shim](https://github.com/vercel/next.js/blob/canary/packages/next/src/lib/detached-promise.ts#L1-L7) for this